vitessio/vitess
# VReplication:  data migration from another Vitess cluster

## Summary

See related RFC for the motivation and design of cross-cluster migration. This PR adds

* a new type of workflow: Migrate in addition to MoveTables and Reshard. 
* a Mount command that manages the CRUD for external clusters in the topo


## Details

## Migrate 
Migrate is mainly a subset of the MoveTables workflow. The parts of the MoveTables implementation which this PR changes	are:


#### Simplified Workflow
Switching traffic is no longer relevant for Migrate workflows since we are not serving these tables in the first place.

#### Source Topo
There are two topologies involved now. The source keyspace belongs to the external cluster, so we need all topo queries to be routed through the topo of the external cluster.

#### Proto 
We add a new attribute to BinlogSource, ExternalCluster, to indicate that the source data resides in an external cluster. Note that this cluster needs to be always mounted in the target Vitess cluster during the entire workflow. 

#### Routing Rules 
While moving tables between keyspaces in a cluster the tables are first served from the source keyspace. Once traffic is switched the tables get served from the target. Vitess uses Routing Rules to ensure that vtgate sees the correct keyspace. However for Migrate, the tables being imported do not exist in the cluster until the import is completed. 

#### VSchema updates
For the same reasons as mentioned above, we only add tables in the target cluster after the Migrate is completed

#### VDiff
VDiff has been changed to use the topo of the external cluster for getting source data.

### Mount

A new sub-key, ExternalCluster, has been added to the global root node of the topo. These will have one child node per type of cluster supported. For now we only support an external vitess cluster. A mysql cluster type will be supported next to point to a regular mysql server. Each cluster is stored under the cluster type with the cluster name and containing the serialized proto as the leaf.

### Testing

We add another cluster and test MoveTables into a "local" cluster from this additional "external" cluster. Since the underlying code is the same, we need to test for two things mainly:
* ensure that the Mount command is storing all the required parameters in the topo. 
* test that MoveTables setups up the workflow correctly for an external topo
* check that all data is copied over correctly using VDiff

Action Item: We need to add the new test to the required CI tests cluster_endtoend_vreplication_migrate.yml after merge

### Test Framework

Currently the VReplication e2e tests work on a single cluster. For cross-cluster tests we need two independent clusters. We had a lot of hard-coded globals to setup ports, temp directories, etc that are the basic parameters needed to setup a cluster. We now allow multiple clusters to be created. 

However: we have an anti-pattern in Vitess: vt executables look for an environment variable VTDATAROOT for certain parameters, like the log directory, path for log files, stdout etc. So each time we need to create vt processes in the "other" cluster we need to set the appropriate VTDATAROOT.

## Related Issue(s)
#7545

## Checklist
- [ ] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

- [ ]  Query Serving
- [X]  VReplication
- [ ]  Cluster Management
- [ ]  VTAdmin
